pkg/sql/logictest/tests/local/local_test: TestLogic_schema_repair failed

pkg/sql/logictest/tests/local/local_test.TestLogic_schema_repair [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/14383219?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/14383219?buildTab=artifacts#/) on master @ [ee5768958fcf7940c73beae06d55377b1b3b3ed8](https://github.com/cockroachdb/cockroach/commits/ee5768958fcf7940c73beae06d55377b1b3b3ed8):


```
=== RUN   TestLogic_schema_repair
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/ab11c9cd60408977be9520945b6e4e92/logTestLogic_schema_repair2224741283
    test_log_scope.go:81: use -show-logs to present logs inline
    logic.go:4308: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/ab11c9cd60408977be9520945b6e4e92/logTestLogic_schema_repair2224741283
--- FAIL: TestLogic_schema_repair (1.69s)
=== RUN   TestLogic_schema_repair/force_delete_data
    logic.go:3011: let $t_id = 108
    logic.go:3011: let $parentID = 104
    logic.go:3011: let $parentSchemaID = 105
    logic.go:3011: let $json = {"table": {"columns": [{"id": 1, "name": "v", "nullable": true, "type": {"family": "IntFamily", "oid": 20, "width": 64}}, {"defaultExpr": "unique_rowid()", "hidden": true, "id": 2, "name": "rowid", "type": {"family": "IntFamily", "oid": 20, "width": 64}}], "createAsOfTime": {}, "families": [{"columnIds": [1, 2], "columnNames": ["v", "rowid"], "defaultColumnId": 1, "name": "primary"}], "formatVersion": 3, "id": 108, "modificationTime": {}, "name": "forcedeletemydata", "nextColumnId": 3, "nextConstraintId": 2, "nextFamilyId": 1, "nextIndexId": 2, "nextMutationId": 1, "parentId": 104, "primaryIndex": {"constraintId": 1, "createdAtNanos": "1710368410157421000", "encodingType": 1, "foreignKey": {}, "geoConfig": {}, "id": 1, "interleave": {}, "keyColumnDirections": ["ASC"], "keyColumnIds": [2], "keyColumnNames": ["rowid"], "name": "forcedeletemydata_pkey", "partitioning": {}, "sharded": {}, "storeColumnIds": [1], "storeColumnNames": ["v"], "unique": true, "version": 4}, "privileges": {"ownerProto": "root", "users": [{"privileges": "2", "userProto": "admin", "withGrantOption": "2"}, {"privileges": "2", "userProto": "root", "withGrantOption": "2"}], "version": 3}, "replacementOf": {"time": {}}, "unexposedParentSchemaId": 105, "version": "1"}}
    logic.go:2965: 
         
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1232/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/schema_repair:190: SELECT * FROM forcedeletemydata ORDER BY v ASC
        expected success, but found
        (XXUUU) unexpected value: raw_bytes:"\257\237s\312\n3\002" timestamp:<wall_time:1710368410165452178 > 
    logic.go:2207: 
         /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1232/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/schema_repair:194: too many errors encountered, skipping the rest of the input
[22:20:10] --- done: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1232/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/schema_repair with config local: 26 tests, 2 failures
--- total: 26 tests, 2 failures
    --- FAIL: TestLogic_schema_repair/force_delete_data (0.11s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogic_schema_repair.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-36691

--- a/pkg/sql/catalog/lease/count.go
+++ b/pkg/sql/catalog/lease/count.go
@@ -15,7 +15,9 @@ import (
 	"fmt"
 	"strings"
 
+	"github.com/cockroachdb/cockroach/pkg/clusterversion"
 	"github.com/cockroachdb/cockroach/pkg/keys"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	clustersettings "github.com/cockroachdb/cockroach/pkg/settings/cluster"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/systemschema"
@@ -43,8 +45,17 @@ func CountLeases(
 	forAnyVersion bool,
 ) (int, error) {
 	// Indicates if the leasing descriptor has been upgraded for session based
-	// leasing.
-	leasingDescIsSessionBased := hasSessionBasedLeasingDesc(ctx, settings)
+	// leasing. Note: Unit tests will never provide cached database regions
+	// so resolve the version from the cluster settings.
+	var systemDBVersion *roachpb.Version
+	if cachedDatabaseRegions != nil {
+		systemDBVersion = cachedDatabaseRegions.GetSystemDatabaseVersion()
+	} else {
+		v := settings.Version.ActiveVersion(ctx).Version
+		systemDBVersion = &v
+	}
+	leasingDescIsSessionBased := systemDBVersion != nil &&
+		systemDBVersion.AtLeast(clusterversion.V24_1_SessionBasedLeasingUpgradeDescriptor.Version())
 	leasingMode := readSessionBasedLeasingMode(ctx, settings)
 	whereClauses := make([][]string, 2)
 	for _, t := range versions {
@@ -61,13 +72,12 @@ func CountLeases(
 				t.ID, versionClause),
 		)
 	}
-
 	whereClauseIdx := make([]int, 0, 2)
 	syntheticDescriptors := make(catalog.Descriptors, 0, 2)
 	if leasingMode != SessionBasedOnly {
-		// The leasing descriptor is not session based yet, so we need to inject
-		// in synthetically.
-		if !leasingDescIsSessionBased {
+		// The leasing descriptor is session based, so we need to inject
+		// expiry based descriptor synthetically.
+		if leasingDescIsSessionBased {
 			syntheticDescriptors = append(syntheticDescriptors, systemschema.LeaseTable_V23_2())
 		} else {
 			syntheticDescriptors = append(syntheticDescriptors, nil)
@@ -76,9 +86,9 @@ func CountLeases(
 
 	}
 	if leasingMode >= SessionBasedDrain {
-		// The leasing descriptor has been upgraded to be session based, so
-		// we need to use a synthetic descriptor for the old expiry based format.
-		if leasingDescIsSessionBased {
+		// The leasing descriptor is not yet session based, so inject the session
+		// based descriptor synthetically.
+		if !leasingDescIsSessionBased {
 			syntheticDescriptors = append(syntheticDescriptors, systemschema.LeaseTable())
 		} else {
 			syntheticDescriptors = append(syntheticDescriptors, nil)

--- a/pkg/sql/catalog/lease/kv_writer.go
+++ b/pkg/sql/catalog/lease/kv_writer.go
@@ -68,7 +68,7 @@ func leaseTableWithID(id descpb.ID, table systemschema.SystemTable) catalog.Tabl
 }
 
 func (w *kvWriter) insertLease(ctx context.Context, txn *kv.Txn, l leaseFields) error {
-	return w.do(ctx, txn, l, func(b *kv.Batch) error {
+	if err := w.do(ctx, txn, l, func(b *kv.Batch) error {
 		// We support writing both session based and expiry based leases within
 		// the KV writer. To be able to support a migration between the two types
 		// of writer will in some cases need to be able to write both types of leases.
@@ -91,11 +91,14 @@ func (w *kvWriter) insertLease(ctx context.Context, txn *kv.Txn, l leaseFields) 
 			}
 		}
 		return nil
-	})
+	}); err != nil {
+		return errors.Wrapf(err, "failed to insert lease %v", l)
+	}
+	return nil
 }
 
 func (w *kvWriter) deleteLease(ctx context.Context, txn *kv.Txn, l leaseFields) error {
-	return w.do(ctx, txn, l, func(b *kv.Batch) error {
+	if err := w.do(ctx, txn, l, func(b *kv.Batch) error {
 		// We support deleting both session based and expiry based leases within
 		// the KV writer. To be able to support a migration between the two types
 		// of writer will in some cases need to be able to delete both types of leases.
@@ -118,7 +121,10 @@ func (w *kvWriter) deleteLease(ctx context.Context, txn *kv.Txn, l leaseFields) 
 			}
 		}
 		return nil
-	})
+	}); err != nil {
+		return errors.Wrapf(err, "failed to delete lease: %v", l)
+	}
+	return nil
 }
 
 type addToBatchFunc = func(*kv.Batch) error

--- a/pkg/sql/catalog/lease/lease.go
+++ b/pkg/sql/catalog/lease/lease.go
@@ -132,10 +132,6 @@ func (m *Manager) sessionBasedLeasingModeAtLeast(
 	return m.getSessionBasedLeasingMode(ctx) >= minimumMode
 }
 
-func hasSessionBasedLeasingDesc(ctx context.Context, settings *cluster.Settings) bool {
-	return !settings.Version.IsActive(ctx, clusterversion.V24_1_SessionBasedLeasingUpgradeDescriptor)
-}
-
 func readSessionBasedLeasingMode(
 	ctx context.Context, settings *cluster.Settings,
 ) SessionBasedLeasingMode {

--- a/pkg/sql/regionliveness/prober.go
+++ b/pkg/sql/regionliveness/prober.go
@@ -114,6 +114,7 @@ type RegionProvider interface {
 type CachedDatabaseRegions interface {
 	IsMultiRegion() bool
 	GetRegionEnumTypeDesc() catalog.RegionEnumTypeDescriptor
+	GetSystemDatabaseVersion() *roachpb.Version
 }
 
 type livenessProber struct {

--- a/pkg/sql/regions/BUILD.bazel
+++ b/pkg/sql/regions/BUILD.bazel
@@ -12,6 +12,7 @@ go_library(
     deps = [
         "//pkg/keys",
         "//pkg/kv",
+        "//pkg/roachpb",
         "//pkg/server/serverpb",
         "//pkg/sql/catalog",
         "//pkg/sql/catalog/descs",

--- a/pkg/sql/regions/cached_db_regions.go
+++ b/pkg/sql/regions/cached_db_regions.go
@@ -15,6 +15,7 @@ import (
 
 	"github.com/cockroachdb/cockroach/pkg/keys"
 	"github.com/cockroachdb/cockroach/pkg/kv"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/lease"
 )
@@ -65,6 +66,11 @@ func (c *CachedDatabaseRegions) IsMultiRegion() bool {
 // database.
 func (c *CachedDatabaseRegions) GetRegionEnumTypeDesc() catalog.RegionEnumTypeDescriptor {
 	return c.dbRegionEnumDesc.AsRegionEnumTypeDescriptor()
+}
+
+// GetSystemDatabaseVersion helper to get the system database version.
+func (c *CachedDatabaseRegions) GetSystemDatabaseVersion() *roachpb.Version {
+	return c.dbDesc.DatabaseDesc().GetSystemDatabaseSchemaVersion()
 }
 
 func TestingModifyRegionEnum(

--- a/pkg/sql/repair.go
+++ b/pkg/sql/repair.go
@@ -37,12 +37,14 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgnotice"
 	"github.com/cockroachdb/cockroach/pkg/sql/privilege"
+	"github.com/cockroachdb/cockroach/pkg/sql/regions"
 	"github.com/cockroachdb/cockroach/pkg/sql/syntheticprivilege"
 	"github.com/cockroachdb/cockroach/pkg/util/duration"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/ioctx"
 	"github.com/cockroachdb/cockroach/pkg/util/log/eventpb"
 	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
+	"github.com/cockroachdb/cockroach/pkg/util/retry"
 	"github.com/cockroachdb/errors"
 )
 
@@ -172,6 +174,22 @@ func (p *planner) UnsafeUpsertDescriptor(
 
 	if force {
 		p.Descriptors().SkipValidationOnWrite()
+	}
+
+	// If we are pushing out a brand new descriptor confirm that no leases
+	// exist before we publish it. This could happen if we did an unsafe delete,
+	// since we will not wait for all leases to expire. So, as a safety force the
+	// unsafe upserts to wait for no leases to exist on this descriptor.
+	if !force &&
+		mut.GetVersion() == 1 {
+		execCfg := p.execCfg
+		regionCache, err := regions.NewCachedDatabaseRegions(ctx, execCfg.DB, execCfg.LeaseManager)
+		if err != nil {
+			return err
+		}
+		if err := execCfg.LeaseManager.WaitForNoVersion(ctx, mut.GetID(), regionCache, retry.Options{}); err != nil {
+			return err
+		}
 	}
 
 	{

--- a/pkg/upgrade/upgrades/v24_1_session_based_lease.go
+++ b/pkg/upgrade/upgrades/v24_1_session_based_lease.go
@@ -128,7 +128,7 @@ func upgradeSystemLeasesDescriptor(
 	ctx context.Context, version clusterversion.ClusterVersion, deps upgrade.TenantDeps,
 ) error {
 	// Upgrade the descriptor in storage to have the new format.
-	return deps.DB.DescsTxn(ctx, func(ctx context.Context, txn descs.Txn) error {
+	if err := deps.DB.DescsTxn(ctx, func(ctx context.Context, txn descs.Txn) error {
 		leaseTable, err := txn.Descriptors().MutableByID(txn.KV()).Table(ctx, keys.LeaseTableID)
 		if err != nil {
 			return err
@@ -146,5 +146,8 @@ func upgradeSystemLeasesDescriptor(
 		leaseTable.NextColumnID = newLeaseTableFormat.TableDescriptor.GetNextColumnID()
 		leaseTable.NextIndexID = newLeaseTableFormat.TableDescriptor.GetNextIndexID()
 		return txn.Descriptors().WriteDesc(ctx, false, leaseTable, txn.KV())
-	})
+	}); err != nil {
+		return err
+	}
+	return bumpSystemDatabaseSchemaVersion(ctx, version, deps)
 }

